src/math/web_mercator.rs

```diff
@@ -0,0 +1,146 @@
+    pub fn to_zoomed_coordinate(&self, z: u8) -> Vector2<f64> {
+        debug_assert!(z <= MAX_ZOOM);
+        // 256 * 2^z
+        let zoom = f64::from(TILE_SIZE << z);
+        zoom * self.xy
+    }
+
+    /// The inverse of [`to_zoomed_coordinate`](#method.to_zoomed_coordinate).
+    pub fn from_zoomed_coordinate(coord: Vector2<f64>, z: u8) -> Self {
+        debug_assert!(z <= MAX_ZOOM);
+        // 256 * 2^z
+        let zoom = f64::from(TILE_SIZE << z);
+        Self { xy: coord / zoom }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use approx::{assert_abs_diff_eq, assert_relative_eq};
+    use nalgebra::Vector2;
+    use nav_types::WGS84;
+
+    /// LAT_BOUND_RAD converted to degrees
+    const LAT_BOUND_DEG: f64 = 85.05112877980658936394320335239172;
+
+    #[test]
+    fn projection_corners() {
+        // Checks that the corners of the map are at the expected coordinates
+        let lat_lng_lower = WGS84::new(LAT_BOUND_DEG, -180.0, 0.0);
+        let lat_lng_upper = WGS84::new(-LAT_BOUND_DEG, 180.0, 0.0);
+        let lower_corner = WebMercatorCoord::from_lat_lng(&lat_lng_lower);
+        let upper_corner = WebMercatorCoord::from_lat_lng(&lat_lng_upper);
+        let lower_corner_truth = Vector2::new(0.0, 0.0);
+        let upper_corner_truth = Vector2::new(256.0, 256.0);
+        assert_abs_diff_eq!(
+            upper_corner.to_zoomed_coordinate(0),
+            upper_corner_truth,
+            epsilon = 10e-10
+        );
+        assert_abs_diff_eq!(
+            lower_corner.to_zoomed_coordinate(0),
+            lower_corner_truth,
+            epsilon = 10e-10
+        );
+    }
+
+    #[test]
+    fn projection_roundtrip() {
+        // Checks that unprojection of a projection returns the original coordinate,
+        // except for altitude, which is 0
+        let test_coordinate = WGS84::new(37.407204, -122.147604, 1300.0);
+        let projected = WebMercatorCoord::from_lat_lng(&test_coordinate);
+        let unprojected = projected.to_lat_lng();
+        assert_relative_eq!(test_coordinate.longitude(), unprojected.longitude());
+        assert_relative_eq!(test_coordinate.latitude(), unprojected.latitude());
+        assert_eq!(unprojected.altitude(), 0.0);
+    }
+
+    #[test]
+    fn projection_ground_truth() {
+        let test_coordinate = WGS84::new(37.407204, -122.147604, 0.0);
+        // This test coordinate is at approx. pixel (165, 18) on this OSM tile at level 19:
+        // https://a.tile.openstreetmap.org/19/84253/203324.png
+        // So the pixel coordinate in a map at zoom level 19 (resolution 256*2^19) is this:
+        let test_coordinate_web_mercator_zoomed_truth =
+            Vector2::new(84253.0 * 256.0 + 165.0, 203324.0 * 256.0 + 18.0);
+        let test_coordinate_web_mercator_zoomed =
+            WebMercatorCoord::from_lat_lng(&test_coordinate).to_zoomed_coordinate(19);
+        // This was from eyeballing, so we shouldn't expect more than 20px accuracy at zoom 19.
```

Comment:
I am readin that level 19 has 0.3m per pixel accuracy at the equator, so 20 pixel would be kind of 6 meters. is it enough for the purpose of this conversion?

Reply:
6 m precision would be too low of course, but that's the precision to which I could measure with this test. I couldn't find a more precise location in the world, since OSM doesn't get deeper than that.

Reply:
if you compared with the results of the common/data_structures/web_mercator of avsw and see if the results are comparable?

Reply:
Yes, they're even exactly the same for this test.